### PR TITLE
Updated wakeup logic to check for spurious wakeups

### PR DESCRIPTION
**Pull request by apgoetz**
Hi all,  

Thanks for all of the great work on this project, I am trying to teach myself Rust by working on a small embedded project, and the work here is super helpful. 

I am currently working on getting the USB suspend/resume logic to work properly in my device, and I think I found a bug in the current code. 

Currently, at least with the 32L0358DISCOVERY Devkit, starting the
stack with the USB port unplugged, or removing the USB port results in
the USB stack being left in state UsbDeviceState::Default. The
expected behavior is that the USB stack stays in the suspended state.

This causes the resume() functions to be called in the HAL, which
means it is not possible to leave the USB hardware in the low power
state.

According to ST document RM0267, section 32.5.5, it is possible to get
spurious wakeup events on the USB bus, even if the bus is still
suspended. Table 159 describes the proper response:

|(RXDP,RXDM)| Wakeup Event        | Required Action         |
|-----------|---------------------|-------------------------|
|  (0,0)    | Root Reset          | None                    |
|  (0,1)    | None (noise on bus) | Go Back in Suspend Mode |
|  (1,0)    | Root Resume         | None                    |
|  (1,1)    | N/A (Noise on bus)  | Go Back in Suspend Mode |

In summary, the status of these bits needs to be checked to make sure
that the source of the wakeup is not spurious noise, and allow the
system to go back to sleep if so.

If you want to test this, I made a small app that printed out the current usb-device bus state to a serial port, and watched for changes as I reset the microcontroller and plugged/unplugged the device. 

I don't have  a standalone demo program, but this repository does include that logic:

https://github.com/apgoetz/led_clock

### CODE REVIEW

**Comment by Disasm** [comment]
Hmm, I think this change is almost the opposite of https://github.com/stm32-rs/stm32-usbd/commit/5fbb1cdca170d7d4aa9ccaa2331ca34de11d8325. I wonder why it works for you.

**Comment by apgoetz** [comment]
I did some more digging into this today, and created a slimmed down example here: https://github.com/apgoetz/usb-suspend

The example program `irq` will print the control and status registers before and after the poll() is called, so that you can see what the poll() ended up doing and what the result is. 

It appears that (at least on my device) the first time FSUSP is written to a 1, an immediate WAKE IRQ is generated. This asynchronously clears the LPMODE bit (so we are no longer in low power mode).

Because of this, I updated my patch to make sure LPMODE is set again if we decide to stay in suspend. 

Looking at the commit above, it seems that someone else had a problem with wake/suspend logic, and the check to RXDP was removed. 

Maybe this is is something that is varying between STM32 microprocessors? or there is something weird going on with my dev board?

**Comment by apgoetz** [comment]
I am closing this since I don't think the changes need to be made to support this issue

**Comment by peku33** [comment]
I've also discovered a bug with spurious wakeups on STM32F070CB. In STM32 reference code, ex. stm32092, they respond to suspend event like this (usbd_pwr.c):
```c
  uint16_t wCNTR;

  /*Store CNTR value */
  wCNTR = _GetCNTR();   
  /* Set FSUSP bit in USB_CNTR register*/
  wCNTR |= CNTR_FSUSP;
  _SetCNTR(wCNTR);
  
  /* force low-power mode in the macrocell */
  wCNTR = _GetCNTR();
  wCNTR |= CNTR_LPMODE;
  _SetCNTR(wCNTR);
```

They perform two write operations, first setting just FSUSP bit, second setting LPMODE, they don't explain why.

This library sets them in one operation: https://github.com/stm32-rs/stm32-usbd/blob/d2196164f3d9d6f9c400e1e7ffb3fca71582445f/src/bus.rs#L291

With such implementation, my device is getting spurious wakeup event. Whenever suspend mode is enabled, it is immediately woken up.

I've changed this to two separate writes, one setting just FRES, second setting LPMODE. The problem seems to be resolved.
